_posts/2018-06-29-node-red-dev-ci.md

```diff
@@ -0,0 +1,590 @@
+* Because Node-RED is built on Node.js you can just run it locally on a lot of systems (Windows, Mac OSX, Linux,...), on a lot of devices and in the cloud.
+
+The graphical user interface lets you create your endpoints and flows in an easy way. Just drag and drop your components in your fowchart and connect them by drawing a line between them. Double clicking each component will open the detail screen where you can set some variables for that component. When your flows are ready, there is import and export functionality available for storing and sharing your configs.
+
+# Why we are using Node-RED (or an alternative)
+
+## Node-RED and Development
+
+Coming from a frontend dev background, I can confirm that nothing is more frustrating then running backends, or at least part of the backend, locally, before you can start coding and testing. Yes, docker and compose are helping a lot but still... In some cases, backend development, for a specific feature, is planned for the same sprint as the frontend development. So even running a backend locally, will not help you. Mocking can be an options. You can start mocking the data inside the communication layer or mocking an external backend with test data.
+
+In an ideal world, teams should do some analysis of the specifications coming from business and their analysts. Based on those specification, developers can start estimating tasks (or stories and epics), and think about the architecture and design. At this point, technical specifications could already get written down, such as model designs and API contracts. It's almost impossible to do a good job in the frontend, if those things are not available, or at least specified.
+
+Why do you want good test data? If you know the characteristics, it's more easy to do some layout stuff. You would think that mockups and designs would be accurate, but most of the time they aren't or they are not inline with the data.
+
+So what about mocking?
+Yes, you can mock stuff yourself and maybe skip the communication layer. Or you can mock the communication part in the communication layer by means of interceptors for example. But at some point in time, you'll need an API and test data that is close to production data when it comes to live data specs. When mocking inside your components, would you include everything in your code base? Or would you exclude it, but still make it more accessible for your colleagues? Some frameworks include this kind of functionality such as MockBackend from Angular (More info <a href="https://angular.io/api/http/testing/MockBackend">here</a>). That said, I really like and prefer external tools for mocking a backend. That way it's separated from your app and code base, so it can easily get replaced or modified without triggering stuff in your frontend pipelines. You can even choose to run a tool on your local machine. Or on premise and make it accessible for all your (frontend) developers.
```

Comment:
Make all links to external resources open in a new tab. We want to keep the reader on our blog.
on-premise